### PR TITLE
London | 25-ITP-SEP | Imran Mohamed | Sprint 2 | Sprint-2 exercises

### PR DESCRIPTION
**Pull request by i786m**
## Learners, PR Template

Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist

Completed the tasks set out in the debug, implement and interpret folders in sprint 2

### CODE REVIEW

File: Sprint-2/debug/recipe.js [hunk 1]
```diff
@@ -12,4 +14,4 @@ const recipe = {
 
 console.log(`${recipe.title} serves ${recipe.serves}
   ingredients:
-${recipe}`);
+${recipe.ingredients}`);
```
Comment (cjyuan) [inline Sprint-2/debug/recipe.js]: This change does not yet satisfy the requirement written on line 6.

File: Sprint-2/implement/lookup.js [hunk 2]
```diff
@@ -1,5 +1,29 @@
...
+        result[country] = currency;
+      }
+      if (currency === null || currency === undefined) {
+        continue;
+      }
+      
+      if (country.length && currency.length && typeof country === 'string' && typeof currency === 'string') {
+        result[country] ? result[country] = [result[country], currency]
+                        : result[country] = currency;
+      }
```
Comment (cjyuan) [inline Sprint-2/implement/lookup.js]: Why check `country.length` and `type of country` on line 20 but not on line 13?

Why check `result[country]` on line 21 but not on line 14?

File: Sprint-2/implement/tally.js [hunk 3]
```diff
@@ -1,3 +1,12 @@
-function tally() {}
+function tally(arr) {
+    if (!Array.isArray(arr) || arguments.length !== 1) {
+        throw new Error("Input must be an array");
+    }
+    const result = {};
+    for (const item of arr) {
+        result[item] = (result[item] || 0) + 1;
+    }
```
Comment (cjyuan) [inline Sprint-2/implement/tally.js]: Does the following function call returns the value you expect?
```
tally(["toString", "toString"]);
```

Suggestion: Look up an approach to create an empty object with no inherited properties.

File: Sprint-2/implement/tally.test.js [hunk 4]
```diff
@@ -19,16 +19,41 @@ const tally = require("./tally.js");
...
+
+//no args passed
+test("should throw an error when no arguments are passed", () => {
+  expect(() => tally()).toThrow("Input must be an array");
+});
+
+//edge cases
+test("should handle array with null and undefined values", () => {
+  expect(tally([null, undefined, null, 'a', undefined])).toEqual({ null: 2, undefined: 2, a: 1 });
+});
```
Comment (cjyuan) [inline Sprint-2/implement/tally.test.js]: What do you expect from the following function call?
```
tally([null, undefined, null, 'a', undefined, "null", "undefined"]
```

File: Sprint-2/implement/lookup.js (hunk 2 above)
Comment (i786m) [reply Sprint-2/implement/lookup.js]: Amended to include checks, and destructure country and currency

File: Sprint-2/implement/lookup.js [hunk 5]
```diff
@@ -1,5 +1,35 @@
...
+         result[country]
+           ? (result[country] = [result[country], currency])
+           : (result[country] = currency);
+      }
+      
+      if (country.length && currency.length && typeof country === 'string' && typeof currency === 'string') {
+         result[country]
+           ? (result[country] = [result[country], currency])
+           : (result[country] = currency);
+      }
```
[the remark this answers is not in the document]
Comment (i786m) [reply Sprint-2/implement/lookup.js]: I have combined both of the conditions so that the checks apply to both strings and arrays for currency, and have refactored the ternary as suggested.
